Allow specifying the Clone Base URL

isotes opened the pull request:
This fixes the detection of the project name for non-standard setups and allows
to specify the preferred URL in the 'Clone Repository' dialog

Fix: #323 #322 #194 #12

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/ui/SettingsPanel.form:
```diff
@@ -22,9 +22,6 @@
               <grid row="0" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="6" anchor="0" fill="1" indent="0" use-parent-layout="false"/>
             </constraints>
             <properties/>
-            <clientProperties>
```
Please revert changes in this file which are not directly related to your change. I know, this is annoying - IntelliJ tries to cleanup things which are required (IMHO).

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/ui/SettingsPanel.form:
```diff
@@ -223,6 +217,35 @@
+                    </constraints>
+                    <properties>
+                      <text value="Clone Base URL"/>
+                      <toolTipText value="Override clone base URL if it differs from gerrit default"/>
```
Set a clone URL if it differs from from Gerrit web URL.

uwolfer commented on src/main/java/com/urswolfer/intellij/plugin/gerrit/ui/SettingsPanel.form:
```diff
@@ -223,6 +217,35 @@
+                      </grid>
+                    </constraints>
+                    <properties>
+                      <toolTipText value="Override clone base URL if it differs from gerrit default"/>
```
This looks duplicated. Is it required for both label and input? Probably you could also add this text as label below this input which would be better visible.

uwolfer reviewed (changes requested):
Thank you very much for this clean PR. I have commented on minor things only.

Can you please describe in a comment what you have tested after implementing this change? It looks backward compatible - have you verified this?

isotes committed:
Allow specifying the Clone Base URL

This fixes the detection of the project name for non-standard setups and allows
to specify the preferred URL in the 'Clone Repository' dialog

Fix: #323 #322 #194 #12

isotes commented:
Thanks for the quick reply and for considering the pull request!

I've made the changes as suggested.

As for backwards compatibility: it was certainly my goal and I think it should work. I've tested the following:
- if the new setting is empty, I have the same behavior as with the existing plugin version
    - the Clone Gerrit dialog shows the same SSH URLs
    - no changes are detected and the Apache server log (reverse proxy setup) shows the query for the wrong project name
- if I set it to the correct (HTTPS) clone URL for our setup
    - the Clone Gerrit dialog shows the correct  HTTPS URLs
    - the correct changes are detected
  
I have no setup that works out of the box with the existing plugin, so I could not check that.

flo-chevalier commented:
Hello @isotes.

I have a question about the "new" CloneBaseUrl field.

In the function GerritGitUtil#getRemoteForChange(), we can see :
`UrlUtils.urlHasSameHost(repositoryUrl, gerritSettings.getHost())`

I have a doubt about : `gerritSettings.getHost()`
If CloneBaseUrl is filled, in that case is never called ?

I would have proposed something like this :
`gerritSettings.getCloneBaseUrlOrHost()`

And in the "GerritSettings" class :
```
public String getCloneBaseUrlOrHost() {
    return cloneBaseUrl == null || cloneBaseUrl.isEmpty() ? host : cloneBaseUrl;
}
```

Maybe I'm wrong.

Thanks for your attention. I’m looking forward to your reply.

Edit:
This is my pull request #349

uwolfer commented:
@isotes: Would be nice if you could review this PR: https://github.com/uwolfer/gerrit-intellij-plugin/pull/349

isotes commented:
Please see my comment on PR #349
